### PR TITLE
feat(sdk): AIP-4946: Update KFP CLI to support recurring runs, archiving experiments and deleting pipeline versions.

### PR DESCRIPTION
**Pull request by alexlatchford**
**Description of your changes:**

Pulls in changes we made to the `kfp` CLI in `aip-kfp-sdk`, but refactors them in a way I believe the KFP folks are more likely to accept as a contribution upstream.

To do:
- [x] Pull in changes from `aip-kfp-sdk` to the `kfp` CLI into `zillow-kfp`. See [this PR](https://gitlab.zgtools.net/analytics/artificial-intelligence/ai-platform/aip-workflow/aip-kfp-sdk/-/merge_requests/43) in `aip-kfp-sdk`.
- Test out changes on:
    - [x] `ci-cd-template`
    - [ ] `~~foundry-ci-cd-template`~~ - Moved to AIP-5426.
- [x] Remove mentions of `aip-kfp-sdk` from the Sphinx docs.

### CODE REVIEW

File: sdk/python/kfp/_client.py [hunk 1]
```diff
@@ -727,6 +743,12 @@ def create_recurring_run(
     Returns:
       A Job object. Most important field is id.
     """
+    if experiment_id is None and experiment_name is None:
+      raise ValueError('Either job_id or experiment_name is required')
+    if not experiment_id:
```
Comment (hsezhiyan) [inline sdk/python/kfp/_client.py:748]: nit: `if` -> `elif`

File: sdk/python/kfp/_client.py [hunk 2]
```diff
@@ -1010,20 +1032,60 @@ def list_recurring_runs(self, page_token='', page_size=10, sort_by='', experimen
       response = self._job_api.list_jobs(page_token=page_token, page_size=page_size, sort_by=sort_by)
     return response
 
-  def get_recurring_run(self, job_id):
+  def get_recurring_run(self, job_id: str = None, job_name: str = None, namespace: str = None):
```
Comment (hsezhiyan) [inline sdk/python/kfp/_client.py:1035]: It's a pity the KFP SDK codebase doesn't have type annotations for outputs. I was going to ask you to write the output annotation, but we can keep as is and keep it consistent.

File: sdk/python/kfp/_client.py [hunk 3]
```diff
@@ -1010,20 +1032,60 @@ def list_recurring_runs(self, page_token='', page_size=10, sort_by='', experimen
...
+    return result.jobs[0]
+
+  def delete_recurring_run(self, job_id: str):
+    """Delete a recurring_run.
 
+    Args:
+      job_id: id of the recurring_run.
+
+    Returns:
+      Object. If the method is called asynchronously, returns the request thread.
```
Comment (hsezhiyan) [inline sdk/python/kfp/_client.py:1083]: Wouldn't the output be a `Job Object`?

**Review by hsezhiyan — APPROVED** [review]

File: sdk/python/kfp/_client.py (hunk 2 above)
Comment (alexlatchford) [reply sdk/python/kfp/_client.py]: It sure does, they just live in the autogenerated `kfp-server-api` library (they basically compile `protobuf` schemas into `python` classes/exceptions that live in that library), I've just updated all the `client` methods with the appropriate return type annotations now, we'll see what `kfp` says during the review! :)

File: sdk/python/kfp/_client.py (hunk 3 above)
Comment (alexlatchford) [reply sdk/python/kfp/_client.py]: Sure would, fixed!

File: sdk/python/kfp/_client.py (hunk 1 above)
Comment (alexlatchford) [reply sdk/python/kfp/_client.py:751]: Yeah I think this is a nuanced one, whilst technically you're correct that both approaches would work I see this as not a valid `elif` case. The rationale is that likely in that case I'd have written the logic instead as:

```python
if not experiment_id:
  if not experiment_name:
    raise ValueError(...)
  
  experiment = self.create_experiment(...)
  ...
```

Whilst this is technically the shortest path I think the code is less clear and consistent across methods in the `client`. Really I like the idea of having the validation standalone as the first step in the method, also this happens to be consistent with how `kfp` have written it too. 

I think this is a classic case of different opinions and all are completely valid and correct, so with that in mind I'm going to resolve with no action just to better conform to the `kfp` standard really so I don't rock the boat too much when I try to get these changes accepted upstream! Thanks for suggestion though 😄

**Pull request merged**